tokio-trace/src/value.rs

```diff
@@ -0,0 +1,363 @@
+/// _or_ may be converted into owned `OwnedValue`s. In addition to being owned,
+/// instances of `OwnedValue` may also be downcast to their original erased type.
+pub trait IntoValue: AsValue {
+    /// Converts this type into an `OwnedValue`.
+    fn into_value(&self) -> OwnedValue;
+}
+
+/// A borrowed value of an erased type.
+///
+/// Like `Any`,`BorrowedValue`s may attempt to downcast the value to
+/// a concrete type. However, unlike `Any`, `BorrowedValue`s are constructed
+/// from types known to implement `AsValue`, providing the `fmt_value` method.
+/// This means that arbitrary `BorrowedValue`s may be formatted using the erased
+/// type's `fmt_value` implementation, _even when the erased type is no longer
+/// known_.
+pub struct BorrowedValue<'a>(&'a dyn AsValue);
+
+/// A `Value` which is formatted using `fmt::Display` rather than `fmt::Debug`.
+#[derive(Clone)]
+pub struct DisplayValue<T: fmt::Display>(T);
+
+/// An owned value of an erased type.
+///
+/// Like `Any`, references to `OwnedValue` may attempt to downcast the value to
+/// a concrete type. However, unlike `Any`, `OwnedValue`s are constructed from
+/// types known to implement `AsValue`, providing the `fmt_value` method.
+/// This means that arbitrary `OwnedValue`s may be formatted using the erased
+/// type's `fmt_value` implementation, _even when the erased type is no longer
+/// known_.
+pub struct OwnedValue(Box<dyn AsValue>);
+
+/// Converts a reference to a `T` into a `BorrowedValue`.
+pub fn borrowed<'a>(t: &'a dyn AsValue) -> BorrowedValue<'a> {
+    BorrowedValue(t)
+}
+
+/// Wraps a type implementing `fmt::Display` so that its `Display`
+/// implementation will be used when formatting it as a `Value`.
+///
+/// # Examples
+/// ```
+/// # extern crate tokio_trace;
+/// use tokio_trace::value;
+/// # use std::fmt;
+/// # fn main() {
+///
+/// #[derive(Clone, Debug)]
+/// struct Foo;
+///
+/// impl fmt::Display for Foo {
+///     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+///         f.pad("Hello, I'm Foo")
+///     }
+/// }
+///
+/// let foo = Foo;
+/// assert_eq!("Foo".to_owned(), format!("{:?}", foo));
+///
+/// let display_foo = value::display(foo.clone());
+/// assert_eq!(
+///     format!("{}", foo),
+///     format!("{:?}", value::borrowed(&display_foo)),
+/// );
+/// # }
+/// ```
+///
+/// ```
+/// # extern crate tokio_trace;
+/// # use std::fmt;
+/// # fn main() {
+/// #
+/// # #[derive(Clone, Debug)]
+/// # struct Foo;
+/// #
+/// # impl fmt::Display for Foo {
+/// #   fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+/// #       f.pad("Hello, I'm Foo")
+/// #   }
+/// # }
+/// use tokio_trace::value::{self, Value, IntoValue};
+/// let foo = value::display(Foo);
+///
+/// let owned_value = foo.into_value();
+/// assert_eq!("Hello, I'm Foo".to_owned(), format!("{:?}", owned_value));
+///
+/// assert!(owned_value.downcast_ref::<Foo>().is_some());
+/// # }
+/// ```
+pub fn display<T>(t: T) -> DisplayValue<T>
```

Comment:
This function is used to select `Display` formatting at the _call-site_ where the span or event is constructed. 

If the _subscriber_ receiving a span or event wants to choose to format something using a formatter other than `Debug`, it will be able to downcast the erased value if it knows it was originally a type that implemented `Display` (or some other formatter).